### PR TITLE
Hashing for Server

### PR DESCRIPTION
**Pull request by WilliamCao13**
*Theoretically works.

### CODE REVIEW

File: src/main/kotlin/com/trackforever/serializer/TrackForeverIssueSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        gen.writeStringField("summary", value.summary)
+        gen.writeObjectField("labels", value.labels)
+        gen.writeObjectField("comments", value.comments)
+        gen.writeStringField("submitterName", value.submitterName)
+        gen.writeObjectField("assignees", value.assignees)
+        if (value.timeCreated != null) {
+            gen.writeNumberField("timeCreated", value.timeCreated!!)
+        } else {
+            gen.writeNumberField("timeCreated", null)
+        }
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/serializer/TrackForeverIssueSerializer.kt]: What's the point of this conditional? Both branches call the same function.

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt [hunk 2]
```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
...
      * Expects a mapping of K-V pairs such that K = projectKey and V = List of Issues as a JSON String
-     * Possible Responses: HTTP 200 OK if all projects exist
-     *                     HTTP 207 MULTI_STATUS with projectKeys mapped to an List of issues that failed to be set.
+     * Possible Responses: HTTP 200 OK with a map of all requested projects to their list of issues
+     *                     HTTP 207 MULTI_STATUS with all projects that could be found mapped to their list of issues
+     *                     HTTP 410 GONE if none of the projects could be found.
      */
     @CrossOrigin
     @PutMapping("/issues")
     fun setIssues(@RequestBody issues: String): ResponseEntity<Map<ProjectId, List<TrackForeverIssue>>> {
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: This should be Map<ProjectId, Map<IssueId, Hash>>

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt [hunk 3]
```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
...
+        // Check with the prevHash
+        // prevHash matches => update that issue for that project
+        // prevHash doesn't match => add that issue to the project
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: If prev hash doesn't match, just skip it

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 2 above)
Comment (cjteam123) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: The Hash should be the new calculated hash for the updated issue

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt [hunk 4]
```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
...
+                        if (it.hash == currentIssue.prevHash) { // there is a match
+                            matchingIssue = it // TODO: some optimization here
+                        }
+                    }
+                    try { // check if the match exists
+                        logger.debug("Preparing to update ${matchingIssue.id} from ${currentProject.id}")
+                        currentProject.issues.remove(matchingIssue.id)
+                    } catch (e: UninitializedPropertyAccessException) { // matchingIssue was never initialized
+                        // No need to delete anything.
+                    }
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: Try using the try catch expression here as you don't do anything with the catch

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt [hunk 5]
```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
...
      * Expects a List of projects in a JSON String.
-     * Possible Responses: HTTP 200 OK upon completion.
-     * TODO: Verify that the projects being passed in are valid
+     * Possible Responses: HTTP 200 OK upon completion with a List of projects
+     *                     that were successfully saved.
      */
     @CrossOrigin
     @PutMapping("/projects")
-    fun setProjects(@RequestBody projects: String): ResponseEntity<String> {
+    fun setProjects(@RequestBody projects: String): ResponseEntity<List<TrackForeverProject>> {
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: Return a List<ProjectId, Hash> here instead.

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt [hunk 6]
```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
...
         val mapper = jacksonObjectMapper()
         val content: List<TrackForeverProject> = mapper.readValue(projects)
+        val successes: MutableList<TrackForeverProject> = mutableListOf()
         // Add to the database
         content.forEach {
-            projectRepository.save(it)
+            // Look in the database for the previous project.
+            val serverProject: TrackForeverProject
+            try { // Project with a prevHash found
+                serverProject = projectRepository.findByHash(it.prevHash)
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: One slight issue with this is that a project can be updated by another user, then the prevHash will never match and a new project will be created instead. You should probably find by project id and then ignore if the prev hash doesn't match. It's the client's duty to merge in remote changes.

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt [hunk 7]
```diff
@@ -96,58 +103,132 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
...
+                        logger.debug("Given hash: ${it.value.hash}")
+                        logger.debug("Expected hash: $calcIssueHash")
+                        it.value.hash = calcIssueHash
+                    }
+                }
+                projectRepository.save(it)
+            } catch (e: EmptyResultDataAccessException) { // No project with such a prevHash found.
+                // Check the newly calculated hash and add it to the repository
+                val calculatedHash = generateHash(it)
+                if (calculatedHash != it.hash) { // Fix hashes for projects
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: This won't work. The hash will be wrong if the user has made changes. The local hash doesn't update until it's been synced with the server.

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt [hunk 8]
```diff
@@ -215,4 +296,34 @@ class ProjectController (@Autowired private var projectRepository: ProjectReposi
...
+                val module = SimpleModule("TrackForeverIssueSerializer", Version(1, 0, 0, null, null, null))
+                module.addSerializer(TrackForeverIssue::class.java, TrackForeverIssueSerializer())
+                mapper.registerModule(module)
+                val issueJson = mapper.writeValueAsString(jsonData)
+                logger.debug("Issue JSON: $issueJson")
+                val crypto = SHA3.Digest512()
+                val issueByteArray = crypto.digest(issueJson.toByteArray())
+                return Hex.toHexString(issueByteArray)
+            }
+            else -> return ""
```
Comment (cjteam123) [inline src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: You can just throw an error here. `throw Error("Should be given a TrackForever Issue or Project")` or something like that

File: src/main/kotlin/com/trackforever/serializer/TrackForeverIssueSerializer.kt (hunk 1 above)
Comment (WilliamCao13) [reply src/main/kotlin/com/trackforever/serializer/TrackForeverIssueSerializer.kt]: Aren't these nullable? I thought I couldn't just do 
```gen.writeNumberField("timeCreated", value.timeCreated)```

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 4 above)
Comment (WilliamCao13) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: I'm a bit confused. Is there a nicer way of doing this in Kotlin?

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 2 above)
Comment (WilliamCao13) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: Are we assuming that IssueIds can never change?

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 5 above)
Comment (WilliamCao13) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: A List or a Map here?

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 2 above)
Comment (cjteam123) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: Yes, we are. There will be an issue with issueid (with multiple clients) clashes, but that's not something we have time to fix.

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 4 above)
Comment (cjteam123) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: You can single-line this and I don't think you need a catch.

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 5 above)
Comment (cjteam123) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: Oh my bad. I mean Map<ProjectId, Hash>
Hash as in the project hash itself - a string.

File: src/main/kotlin/com/trackforever/serializer/TrackForeverIssueSerializer.kt (hunk 1 above)
Comment (cjteam123) [reply src/main/kotlin/com/trackforever/serializer/TrackForeverIssueSerializer.kt]: You should be able to!

File: src/main/kotlin/com/trackforever/controllers/ProjectController.kt (hunk 4 above)
Comment (WilliamCao13) [reply src/main/kotlin/com/trackforever/controllers/ProjectController.kt]: I tried removing the catch, but it still throws an error. I think I'll just single-line the thing.

**Pull request merged**